kong/clustering/rpc/manager.lua

```diff
@@ -0,0 +1,288 @@
+local _M = {}
+local _MT = { __index = _M, }
+
+
+local server = require("resty.websocket.server")
+local client = require("resty.websocket.client")
+local socket = require("kong.clustering.rpc.socket")
+local concentrator = require("kong.clustering.rpc.concentrator")
+local future = require("kong.clustering.rpc.future")
+local utils = require("kong.clustering.rpc.utils")
+local callbacks = require("kong.clustering.rpc.callbacks")
+local clustering_tls = require("kong.clustering.tls")
+local constants = require("kong.constants")
+local table_isempty = require("table.isempty")
+local pl_tablex = require("pl.tablex")
+
+
+local ngx_var = ngx.var
+local ngx_ERR = ngx.ERR
+local ngx_log = ngx.log
+local ngx_exit = ngx.exit
+local exiting = ngx.worker.exiting
+local pl_tablex_makeset = pl_tablex.makeset
+local validate_client_cert = clustering_tls.validate_client_cert
+
+
+local WS_OPTS = {
+  timeout = constants.CLUSTERING_TIMEOUT,
+  max_payload_len = kong.configuration.cluster_max_payload,
+}
+local KONG_VERSION = kong.version
+
+
+-- create a new RPC manager, node_id is own node_id
+function _M.new(conf, node_id)
+  local self = {
+    -- clients[node_id]: { socket1 => true, socket2 => true, ... }
+    clients = {},
+    client_capabilities = {},
+    node_id = node_id,
+    conf = conf,
+    cluster_cert = assert(clustering_tls.get_cluster_cert(conf)),
+    cluster_cert_key = assert(clustering_tls.get_cluster_cert_key(conf)),
+    callbacks = callbacks.new(),
+  }
+
+  self.concentrator = concentrator.new(self, kong.db)
+
+  self.callbacks:register("kong.meta.v1.capability_advertisement", function(node_id, capabilities)
```

Comment:
should we define this `"kong.meta.v1.capability_advertisement"` to a constant?

Reply:
Using plain strings here does simplify the code review process. I believe code readability is more important than the beauty of code structure. The code should be designed for human comprehension. 😸

Reply:
It's appropriate to allow some constants to be hardcoded directly into the code.